packages/metadata/src/decorator-factory.ts

```diff
@@ -611,9 +638,34 @@ export class MethodParameterDecoratorFactory<T> extends DecoratorFactory<
     methodDescriptor?: TypedPropertyDescriptor<any> | number,
   ) {
     ownMetadata = ownMetadata || {};
-    let params = ownMetadata[methodName!];
-    params = [this.spec].concat(params);
+    const numOfParams = this.getNumberOfParameters(target, methodName);
+    // Fetch the cached parameter index
+    let index = Reflector.getOwnMetadata(
+      this.key + ':index',
+      target,
+      methodName,
+    );
+    // Default to the last parameter
+    if (index == null) index = numOfParams - 1;
+    if (index < 0) {
+      // Excessive decorations than the number of parameters detected
+      const method = this.getTargetName(target, methodName, methodDescriptor);
+      throw new Error(
+        `The decorator is used more than ${numOfParams} time(s) on ${method}`,
+      );
+    }
```

Comment:
You've got the same logic for the index checking and error throwing in both places; I'd extract it into a function.